vercel/turborepo
# eslint plugin no-undeclared-env-vars should detect known framework variables

### Description

This PR picks up from https://github.com/vercel/turborepo/pull/8505

We currently do [Framework Inference](https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#framework-inference) based on the package.json dependencies of the nearest package.json. turbo's Framework Inference support is per-package whereas this one wholesale allowslists all framework environment variables in every package.

The job of this PR, therefore, is to enable per-package framework detection.

I started by extracting this data to a json file, frameworks.json, so it can be shared between Rust and TypeScript.

TODOs:
- [x] extract the frameworks information in `framework.rs` into `frameworks.json`
- [x] search community to make sure that recursive package.json searching has no better method than fs.readFileSync
- [x] implement package.json search to resolve depencies
- [x] combine per-package resolution of those dependencies and regex allowlist values
- [ ] update tests for the plugin
- [ ] use the `frameworks.json` file to generate the table in `using-environment-variables.mdx`

### Testing Instructions

After building the plugin (which will run it locally), open up `packages/eslint-plugin-turbo/__fixtures__/framework-inference/apps/vite`.  Notice that there are no errors in `index.js`, but if you remove the `vite` dependency in `package.json` then there will be.  Notice, also, that adding a next.js env var in this file (e.g. `process.env.NEXT_PUBLIC_ZILTOID`) will error as expected.
